### ISSUE TITLE
Tested remote debugging with Termdebug: a small report and a feature request. 

### ISSUE THREAD

**Issue by ubaldot** [vim/vim#15293]
I tried to remote debugging with Termdebug, and everything worked great.  with the exception of a small singularity on Windows.

The setup was  `gdb `connected to an `openocd `server connected to a _stlink_ probe and a _stm32 Nucleo_ board. 

With this setup in Windows, when you run `continue ` from the gdb console and there are no breakpoints, the control over the gdb console has gone and it seems that you cannot get it back. 

Typically,  to get the control back you hit CTRL_C on the gdb console, but in this case nothing happens. Furthermore, if you try to type commands in the gdb console, you won't get any message back. 
Nevertheless, and somehow surprisingly, if you run  `<c-w>:q` from the gdb console the messages that you were supposed to get back in response to your commands are displayed all at once.

The following is what I see after after I run `continue`

```
gdb> continue
Continuing.
gdb> monitor halt
gdb> monitor halt
gdb>
```

and the following is what I get displayed when I run `<c-w>:q`

```
gdb> continue
Continuing.
gdb> monitor halt
gdb> monitor halt
"monitor" command not supported by this target.
"monitor" command not supported by this target.
```

The command `monitor halt` was perfectly working before running `continue`. 

I tried the same experiment on macos in Termdebug prompt mode and hitting `CTRL_C` after a `continue` command gave me back the gdb control, as it is supposed to be.

I analyzed the gdb logs of both Windows and macos to check the communication between gdb and openocd and it appear to identical:

```
...
&"c\n"
~"Continuing.\n"
^running
*running,thread-id="all"
(gdb) 
&"[remote] Sending packet: $vCont?#49\n"
&"[remote] Packet received: vCont;c;C;s;S\n"
&"[remote] packet_ok: Packet vCont (verbose-resume) is supported\n"
&"[remote] Sending packet: $vCont;c#a8\n"
&"[remote] wait: enter\n"
&"[remote] wait: exit\n"
&"[remote] wait: enter\n"
&"  [remote] Packet received: O\n"
&"[remote] wait: exit\n"
&"[remote] wait: enter\n"
&"  [remote] Packet received: O\n"
&"[remote] wait: exit\n"
&"[remote] wait: enter\n"
&"  [remote] Packet received: O\n"
&"[remote] wait: exit\n"
&"[remote] wait: enter\n"
&"  [remote] Packet received: O\n"
&"[remote] wait: exit\n"
&"[remote] wait: enter\n"
...
```

Hence, I would exclude possible communication problems between gdb and openocd. 

However, by looking at the Termdebug source code, I noticed that there is a special function called `debugbreak()` which seems to be designed to take back the control of the gdb console when using Windows. 
Therefore, I tried to get both the gdb and openocd PIDs from a powershell (`Get-Process -Name {arm-none-eabi-gdb, openocd} | Select-Object -Property Id`) and run `echo debugbreak(pid_openocd)` first and ` echo debugbreak(pid_gdb)`  in another working session returning `1` in both cases - in my understanding you should only use the `debugbreak(open_ocd)` but I tried both anyways. . Yet, I couldn't get control back of the gdb console. 

I am suspicious that the `debugbreak()` function works well when you have a "debugged program" window in the classical sense,  but for remote debbuging, or at least in this use-case. Hence, if this is the issue, my enhancement request would be to enhance  `debugbreak()`  to further deal with remote debugging, if possible. 

In conclusion, using Termdebug for remote debugging seems to work fine, as long as one avoid to run `continue` when there are no breakpoints aand when working with Windows. 

Sorry for the essay. :)

**Comment by ubaldot** [comment]
Possible workaround (on the user-side and not tested):

Download SendSignalCtrlC.exe ([download link](https://github.com/SergeyPirogov/video-recorder-java/raw/master/core/src/main/resources/SendSignalCtrlC.exe)) and from powershell run SendSignalCtrlC.exe 12345, where 12345 is the gdb PID.

You could have something like:

if executable('SendSignalCtrlC') && has('win32')
  var gdb_pid =  system($'powershell Get-Process -Name {gdb_bufname} | ForEach-Object { $_.Id }') 
  exe $"!SendSignalCtrlC.exe {gdb_pid}"
endif 

to mimic CTRL_C on the gdb process.

**Comment by errael** [comment]
BTW, are you familiar with the `ch_log` stuff. Just wondering if `termdebug`'s log, in addition to `gdb`'s might be useful.

**Comment by ubaldot** [comment]
> BTW, are you familiar with the `ch_log` stuff. Just wondering if `termdebug`'s log, in addition to `gdb`'s might be useful.

Point. Yes, I run again and I sent few commands after `continue`. I filtered the `debuglog` file created by  `ch_logfile('debuglog', 'w')` with `:g/ch_log()` getting the following (excerpt after `continue `command) :  


```
2875  16.162011 : ch_log(): sending to gdb: continue
2895  16.180201 : ch_log(): received from gdb: &"continue\n"
2897  16.180229 : ch_log(): received from gdb: ~"Continuing.\n"
2899  16.180320 : ch_log(): received from gdb: ^running
2901  16.180415 : ch_log(): received from gdb: *running,thread-id="all"
2902  16.180437 : ch_log(): program running
2904  16.180469 : ch_log(): received from gdb: (gdb) 
2906  16.180482 : ch_log(): received from gdb: &"[remote] Sending packet: $vCont?#49\n"
2908  16.180499 : ch_log(): received from gdb: &"[remote] Packet received: vCont;c;C;s;S\n"
2910  16.180515 : ch_log(): received from gdb: &"[remote] packet_ok: Packet vCont (verbose-resume) is supported\n"
2912  16.180530 : ch_log(): received from gdb: &"[remote] Sending packet: $vCont;c#a8\n"
2914  16.180546 : ch_log(): received from gdb: &"[remote] wait: enter\n"
2916  16.180560 : ch_log(): received from gdb: &"[remote] wait: exit\n"
2968  16.695998 : ch_log(): received from gdb: &"[remote] wait: enter\n"
2970  16.696043 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
2972  16.696065 : ch_log(): received from gdb: &"[remote] wait: exit\n"
2988  17.205191 : ch_log(): received from gdb: &"[remote] wait: enter\n"
2990  17.205212 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
2992  17.205221 : ch_log(): received from gdb: &"[remote] wait: exit\n"
3009  17.720788 : ch_log(): received from gdb: &"[remote] wait: enter\n"
3011  17.720833 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
3013  17.720855 : ch_log(): received from gdb: &"[remote] wait: exit\n"
3033  18.231390 : ch_log(): received from gdb: &"[remote] wait: enter\n"
....
3724  29.321112 : ch_log(): sending to gdb: monitor halt
3738  29.491754 : ch_log(): received from gdb: &"[remote] wait: enter\n"
3740  29.491776 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
3742  29.491786 : ch_log(): received from gdb: &"[remote] wait: exit\n"
3758  30.003945 : ch_log(): received from gdb: &"[remote] wait: enter\n"
3760  30.003984 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
3762  30.004013 : ch_log(): received from gdb: &"[remote] wait: exit\n"
3790  30.519453 : ch_log(): received from gdb: &"[remote] wait: enter\n"
3792  30.519470 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
3794  30.519478 : ch_log(): received from gdb: &"[remote] wait: exit\n"
3829  31.028579 : ch_log(): received from gdb: &"[remote] wait: enter\n"
3831  31.028616 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
3833  31.028648 : ch_log(): received from gdb: &"[remote] wait: exit\n"
3867  31.540363 : ch_log(): received from gdb: &"[remote] wait: enter\n"
....
4158  36.178463 : ch_log(): sending to gdb: interrupt
4172  36.658684 : ch_log(): received from gdb: &"[remote] wait: enter\n"
4174  36.658727 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
4176  36.658765 : ch_log(): received from gdb: &"[remote] wait: exit\n"
4204  37.173201 : ch_log(): received from gdb: &"[remote] wait: enter\n"
4206  37.173229 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
4208  37.173249 : ch_log(): received from gdb: &"[remote] wait: exit\n"
4242  37.686056 : ch_log(): received from gdb: &"[remote] wait: enter\n"
4244  37.686093 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
4246  37.686111 : ch_log(): received from gdb: &"[remote] wait: exit\n"
4278  38.198646 : ch_log(): received from gdb: &"[remote] wait: enter\n"
4280  38.198677 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
4282  38.198695 : ch_log(): received from gdb: &"[remote] wait: exit\n"
4308  38.712366 : ch_log(): received from gdb: &"[remote] wait: enter\n"
...

... a more detail log excerpt follows

**Comment by ubaldot** [comment]
A more detailed log excerpt after `continue`:

 ```
16.162011 : ch_log(): sending to gdb: continue
 16.162026 SEND on 6(in): 'continue
'
 16.162331 : SafeState: Start triggering
 16.163168 : looking for messages on channels
 16.163183 : SafeState: back to waiting, triggering SafeStateAgain
 16.180145 RECV on 6(out): '&"continue\n"

~"Continuing.\n"

^running

*running,thread-id="all"

(gdb) 

&"[remote] Sending packet: $vCont?#49\n"

&"[remote] Packet received: vCont;c;C;s;S\n"

&"[remote] packet_ok: Packet vCont (verbose-resume) is supported\n"

&"[remote] Sending packet: $vCont;c#a8\n"

&"[remote] wait: enter\n"

&"[remote] wait: exit\n"

'
 16.180174 RECV on 5(out): '[?25lDeb[0K[1G[?25h'
 16.180186 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180201 : ch_log(): received from gdb: &"continue\n"
 16.180220 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180229 : ch_log(): received from gdb: ~"Continuing.\n"
 16.180305 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180320 : ch_log(): received from gdb: ^running
 16.180404 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180415 : ch_log(): received from gdb: *running,thread-id="all"
 16.180437 : ch_log(): program running
 16.180459 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180469 : ch_log(): received from gdb: (gdb) 
 16.180475 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180482 : ch_log(): received from gdb: &"[remote] Sending packet: $vCont?#49\n"
 16.180492 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180499 : ch_log(): received from gdb: &"[remote] Packet received: vCont;c;C;s;S\n"
 16.180508 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180515 : ch_log(): received from gdb: &"[remote] packet_ok: Packet vCont (verbose-resume) is supported\n"
 16.180523 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180530 : ch_log(): received from gdb: &"[remote] Sending packet: $vCont;c#a8\n"
 16.180539 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180546 : ch_log(): received from gdb: &"[remote] wait: enter\n"
 16.180554 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.180560 : ch_log(): received from gdb: &"[remote] wait: exit\n"
 16.180569 on 5: writing 23 bytes to terminal
 16.180582 on 5: updating screen
 16.182463 : SafeState: back to waiting, triggering SafeStateAgain
 16.195817 : looking for messages on channels
 16.195835 : SafeState: back to waiting, triggering SafeStateAgain
 16.211375 RECV on 5(out): 'Debug: 1036 17153 gdb_server.c:375 gdb_log_incoming_packet([0K)[?25l

: [stm32f4x.cpu] received packet: vCont?[0K

Debug: 1037 17153 gdb_server.c:391 gdb_log_outgoing_packet([0K)

: [stm32f4x.cpu] sending packet: $vCont;c;C;s;S#62[0K

Debug: 1038 17153 gdb_server.c:375 gdb_log_incoming_packet([0K)

: [stm32f4x.cpu] received packet: vCont;c[0K

Debug: 1039 17153 gdb_server.c:3049 gdb_handle_vcont_packet[0K(

): target stm32f4x.cpu continue[0K

Debug: 1040 17168 target.c:1815 target_call_event_callbacks[0K(

): target event 3 (resume-start) for core stm32f4x.cpu[0K

Debug: 1041 17168 hla_target.c:448 adapter_resume(): adapte[0Kr

_resume 1 0x00000000 0 0[0K

Debug: 1042 17168 target.c:2171 target_free_all_working_are[0Ka

s_restore(): freeing all working area[0K[35G[?25h'
 16.211388 on 5: writing 796 bytes to terminal
 16.211512 on 5: updating screen
 16.214353 : SafeState: back to waiting, triggering SafeStateAgain
 16.216088 : looking for messages on channels
 16.216098 : SafeState: back to waiting, triggering SafeStateAgain
 16.242656 RECV on 5(out): '[?25l
s_restore(): freeing all working areas[0K

Debug: 1043 17184 target.c:2688 target_write_u32(): address[0K:

 0xe000edfc, value: 0x01000000[0K

Debug: 1044 17184 hla_target.c:616 adapter_write_memory(): [0Ka

dapter_write_memory 0xe000edfc 4 1[0K

Debug: 1045 17184 armv7m.c:174 armv7m_restore_context():[0K

Debug: 1046 17184 target.c:2688 target_write_u32(): address[0K:

 0xe000edf8, value: 0x00000000[0K

Debug: 1047 17184 hla_target.c:616 adapter_write_memory(): [0Ka

dapter_write_memory 0xe000edf8 4 1[0K

Debug: 1048 17200 target.c:1815 target_call_event_callbacks[0K(

): target event 2 (resumed) for core stm32f4x.cpu[0K

Debug: 1049 17200 target.c:1815 target_call_event_callbacks[0K(

): target event 4 (resume-end) for core stm32f4x.cpu[0K

Debug: 1050 17200 target.c:1815 target_call_event_callbacks[0K(

): target event 7 (gdb-start) for core stm32f4x.cpu[0K

[0K[?25h'
 16.242687 on 5: writing 907 bytes to terminal
 16.242852 on 5: updating screen
 16.246454 : SafeState: back to waiting, triggering SafeStateAgain
 16.258430 : looking for messages on channels
 16.258484 : SafeState: back to waiting, triggering SafeStateAgain
 16.695929 RECV on 6(out): '&"[remote] wait: enter\n"

&"  [remote] Packet received: O\n"

&"[remote] wait: exit\n"

'
 16.695973 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.695998 : ch_log(): received from gdb: &"[remote] wait: enter\n"
 16.696028 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.696043 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
 16.696057 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 16.696065 : ch_log(): received from gdb: &"[remote] wait: exit\n"
 16.696106 : SafeState: back to waiting, triggering SafeStateAgain
 16.711681 RECV on 5(out): 'Debug: 1051 17668 gdb_server.c:391 gdb_log_outgoing_packet([0K)[?25l

: [stm32f4x.cpu] sending packet: $O#4f[0K

[0K[?25h'
 16.711705 : looking for messages on channels
 16.711713 on 5: writing 126 bytes to terminal
 16.711787 on 5: updating screen
 16.714545 : SafeState: back to waiting, triggering SafeStateAgain
 16.727337 : looking for messages on channels
 16.727384 : SafeState: back to waiting, triggering SafeStateAgain
 17.205110 RECV on 6(out): '&"[remote] wait: enter\n"

&"  [remote] Packet received: O\n"

&"[remote] wait: exit\n"

'
 17.205163 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 17.205191 : ch_log(): received from gdb: &"[remote] wait: enter\n"
 17.205208 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 17.205212 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
 17.205217 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 17.205221 : ch_log(): received from gdb: &"[remote] wait: exit\n"
 17.205247 : SafeState: back to waiting, triggering SafeStateAgain
 17.220700 RECV on 5(out): 'Debug: 1052 18178 gdb_server.c:391 gdb_log_outgoing_packet([0K)[?25l

: [stm32f4x.cpu] sending packet: $O#4f[0K

[0K[?25h'
 17.220725 : looking for messages on channels
 17.220733 on 5: writing 126 bytes to terminal
 17.220768 on 5: updating screen
 17.222504 : SafeState: back to waiting, triggering SafeStateAgain
 17.236296 : looking for messages on channels
 17.236312 : SafeState: back to waiting, triggering SafeStateAgain
 17.720706 RECV on 6(out): '&"[remote] wait: enter\n"

&"  [remote] Packet received: O\n"

&"[remote] wait: exit\n"

'
 17.720756 RECV on 5(out): '[?25lDebug[0K[3G[?25h'
 17.720767 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 17.720788 : ch_log(): received from gdb: &"[remote] wait: enter\n"
 17.720823 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 17.720833 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
 17.720847 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 17.720855 : ch_log(): received from gdb: &"[remote] wait: exit\n"
 17.720865 on 5: writing 25 bytes to terminal
 17.720879 on 5: updating screen
 17.721118 : SafeState: back to waiting, triggering SafeStateAgain
 17.736408 : looking for messages on channels
 17.736440 : SafeState: back to waiting, triggering SafeStateAgain
 17.751943 RECV on 5(out): '[?25l
Debug: 1053 18693 gdb_server.c:391 gdb_log_outgoing_packet([0K)

: [stm32f4x.cpu] sending packet: $O#4f[0K

[0K[?25h'
 17.751997 on 5: writing 127 bytes to terminal
 17.752061 on 5: updating screen
 17.754458 : SafeState: back to waiting, triggering SafeStateAgain
 17.767580 : looking for messages on channels
 17.767612 : SafeState: back to waiting, triggering SafeStateAgain
 18.231309 RECV on 6(out): '&"[remote] wait: enter\n"

&"  [remote] Packet received: O\n"

&"[remote] wait: exit\n"

'
 18.231356 RECV on 5(out): '[?25lDebug:[0K[2G[?25h'
 18.231367 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 18.231390 : ch_log(): received from gdb: &"[remote] wait: enter\n"
 18.231419 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 18.231427 : ch_log(): received from gdb: &"  [remote] Packet received: O\n"
 18.231437 on 6: Invoking channel callback <SNR>166_GdbOutCallback
 18.231444 : ch_log(): received from gdb: &"[remote] wait: exit\n"
 18.231454 on 5: writing 26 bytes to terminal
 18.231468 on 5: updating screen
 18.231693 : SafeState: back to waiting, triggering SafeStateAgain
 18.247049 : looking for messages on channels
 18.247087 : SafeState: back to waiting, triggering SafeStateAgain
 18.262586 RECV on 5(out): '[?25l
Debug: 1054 19204 gdb_server.c:391 gdb_log_outgoing_packet([0K)

: [stm32f4x.cpu] sending packet: $O#4f[0K

[0K[?25h'

```
